rfd/0051-x11-forwarding.md

```diff
@@ -0,0 +1,205 @@
+server/client and Teleport server/client.
+
+## Details
+
+X11 forwarding is a sparsely documented protocol with many hidden intricacies.
+In implementing X11 forwarding for Teleport I leaned heavily upon OpenSSH
+implementation, which is (was) the only complete open source X11 forwarding
+implementation.
+
+This RFD will explain some key concepts and temrs for X11 forwarding, the general
+flow of the implementation, and the security details and decisions.
+
+### X Server and Client
+
+An X Server is an application on a machine which manages graphical displays, input
+devices, etc. Linux systems ship with an XServer (XOrg), while other operating
+systems require a standalone XServer (ex. XQuartz for Mac, XMing for Windows).
+
+An X Client is a application which interacts with an X server to send/receive
+graphical input and output. For Max and Linux, most terminals have an X Client
+built in. On Windows, you have to install a standalone X Client such as MobaXTerm.
+
+When an X Client makes an X request, it parses the set `$DISPLAY` to send an X request
+to a specific display opened in the X Server. the first portion of this request is
+an authorization packet used to authenticate and authorize the X Client.
+
+The X Client gets the authorization packet from `$XAUTHORITY` or `~/.Xauthority`.
+If there is no matching authorization data in the file, or it doesn't match what
+the X Server expected, then the X request will be denied.
+
+### X11 Forwarding flow
+
+1. Client opens an SSH session and then sends an `x11-req` ssh request to the server. 
+   The `x11-req` includes an X authentication protocol and cookie, which will be detailed
+   in the security sections below.
+2. Server receives request and either approves or denies the request.
+3. If the server approves, it will open an XServer proxy listener to catch any
+   XServer requests within the SSH session.
+4. Upon receiving an XServer request on the proxy listener, the server sends an
+   `x11` channel request to the client.
+5. If the client accepts, the client will start X11 forwarding between the x11
+   channel and the client's local display - `$DISPLAY`.
+6. Server begins X11 forwarding between the x11 channel and XServer proxy connection.
+7. The XServer request is now connected to the client's local display. For example, if
+   `xeyes` was run, then the program would be running on the server, but any graphical
+   input/output would the client's local XServer.
+
+### X Server Proxy
+
+In step 3 above, the SSH server opens an X Server Proxy. More specifically, the server
+starts listening on a unix socket so that it can accept requests to the corresponding
+display. 
+
+First, the server has to find an open display socket. A display's unix socket can be 
+determined by its display number - `/tmp/.X11-unix/X<display_number>`. So we start
+searching for an open display number between 10 and 1009.
+
+Note: We use the same display offset (10) and max displays (1000) as OpenSSH.
```

Comment:
Why is an offset necessary? Why not start looking at display 0?

Reply:
To avoid overlapping with local displays which start counting up from 0. It's not strictly necessary, but for consistency/clarity we add the offset to separate local and proxy displays. This way `ls /tmp/.X11-unix` might output `X0 X1 X2 X10 X11 X12 X13`, and it's clear where each unix socket originated. Adding a shorter explanation to the RFD.